### DISCUSSION TITLE
Filter by key of array of objects

### DISCUSSION THREAD

**Discussion by brydonm** [supabase/supabase#4784] [Questions] [answered]
Hey,

I'm trying to query my table based on the value of an object with an array of objects.

My `wars` table looks like this:
|id|company|roster|
|---|---|---|
|1|"Big Coin"|[{"id": "abcd"}, {"id": "efgh"}]|

I'm trying to fetch wars from the `supabase-js` library that include that user in the roster.

This is the query that I've been playing with but can't seem to get it to work.
```js
supabase
      .from("wars")
      .select(`
        id,
        ...
      `)
      .contains("roster", ["roster->>id: 'abcd'"]); // Should return id 1 from `wars`, along with any other wars this player is a part of
```

I've been trying to use [this documentation](https://supabase.com/docs/reference/javascript/contains) on `contains` but it doesn't have a lot of info.

Any help would be appreciated, thank you!

**Comment by steve-chavez** [comment, marked as answer]
> .contains("roster", ["roster->>id: 'abcd'"]); 

The above is matching to a value of `"roster->>id: 'abcd'"`(doesn't get translated to an expression in any way).

What's the type of the `roster` column? If it's a `jsonb` then you should be able to do:

```js
supabase
      .from("wars")
      .select(`
        id,
        ...
      `)
      .contains("roster", [{"id": "abcd"}]); 
```

On SQL, an equivalent comparison would be:

```sql
select '[{"id": "abcd"}, {"id": "efgh"}]'::jsonb @> '[{"id": "abcd"}]';
```

**Reply by brydonm** [reply]
@steve-chavez  it was `json` but I swapped to `jsonb` to test this - I'm getting this error:

```
code: "22P02"
details: "Token \"object\" is invalid."
hint: null
message: "invalid input syntax for type json"
```

```js
supabase
      .from("wars")
      .select(
        `
      id,
      starts_at,
      discord_link,
      won,
      company,
      zone,
      war_code,
      roster,
      commanders
  `
      )
      .contains("roster", [{ id: user.id }]);
```

thank you for your help, again :)

edit: looks like that was being passed as `{[Object object]}` trying to convert it to a string with a few methods (one of which below) doesn't seem to be working for other various errors (`message: "malformed array literal: \"[{ id: 'cd7961ec-a9ea-4e50-9889-19a66646631b' }]\""`)

```
.contains("roster", `[{ id: '${user.id}' }]`);
```

**Reply by steve-chavez** [reply]
Can you try with:

```js
 .contains("roster", JSON.stringify([{ id: user.id }]));
```

**Reply by brydonm** [reply]
@steve-chavez that is one of the methods I tried but unfortunately still no work :(

```json
{
  "message": "malformed array literal: \"[{ 'id': 'cd7961ec-a9ea-4e50-9889-19a66646631b' }]\"",
  "code": "22P02",
  "hint": null,
  "details": "\"[\" must introduce explicitly-specified array dimensions."
}
```

```
select: id,starts_at,discord_link,won,company,zone,war_code,roster,commanders
roster: cs.[{ 'id': 'cd7961ec-a9ea-4e50-9889-19a66646631b' }]
```

**Reply by steve-chavez** [reply]
Is your type defined as `jsonb[]` or `jsonb`?

**Reply by brydonm** [reply]
ah that worked! it was `jsonb[]`. I switched it to `json` and went back to this method: `.contains("roster", JSON.stringify([{ id: user.id }]));` and it worked.

Thanks so much for your help Steve!

**Reply by Kikky** [reply]
> Can you try with:
> 
> ```js
>  .contains("roster", JSON.stringify([{ id: user.id }]));
> ```

This helped me. Thanks!

**Reply by Strijdhagen** [reply]
> ah that worked! it was `jsonb[]`. I switched it to `json` and went back to this method: `.contains("roster", JSON.stringify([{ id: user.id }]));` and it worked.
> 
> Thanks so much for your help Steve!

For anyone that googled this and still cant figure out why you cant filter your JSONB[]. It's the [] that is the issue, not JSON or JSONB.
Changing my view to JSONB columns instead of JSON[] fixed the issues for me.
